### PR TITLE
refactor(flow): func spec api&use Error not EvalError in mfp

### PR DESCRIPTION
**Pull request by discord9**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

__!!! DO NOT LEAVE THIS BLOCK EMPTY !!!__

Please explain IN DETAIL what the changes are in this PR and why they are needed:

Make some changes to various types to make them easy to use.

- Summarize your change (**mandatory**)
    - refactor the functions for `*Function` to choose what specialized function to use when input types are known.
    - Also allow `signature` to have multiple input and refine defining signatures with macro
    - Use `Error` for `MapFilterProject` 's method that should only be called when building a dataflow graph
    - `map`, `projection` and `filter` for `TypedPlan` since they are commonly used in building those plan from substrait
    - some new methods for ColumnType
## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/flow/clippy.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Whether to only check for missing documentation in items visible within the current crate. For example, pub(crate) items. (default: false)
+# This is a config for clippy::missing_docs_in_private_items
+missing-docs-in-crate-items = true
```
Comment (zhongzc) [inline src/flow/clippy.toml:3]:
```suggestion
missing-docs-in-crate-items = true

```

File: src/flow/src/expr/func.rs [hunk 2]
```diff
@@ -47,6 +82,60 @@ pub enum UnaryFunc {
...
+            "is_true" => Ok(Self::IsTrue),
+            "is_false" => Ok(Self::IsFalse),
+            "step_timestamp" => Ok(Self::StepTimestamp),
+            "cast" => {
+                let arg_type = arg_type.ok_or_else(|| {
+                    InvalidQuerySnafu {
+                        reason: "cast function requires a type argument".to_string(),
+                    }
+                    .build()
+                })?;
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:138]: `with_context` instead?

File: src/flow/src/expr/func.rs [hunk 3]
```diff
@@ -154,7 +243,198 @@ pub enum BinaryFunc {
...
+            .cloned()
+            .ok_or_else(|| {
+                InvalidQuerySnafu {
+                    reason: format!(
+                        "No specialization found for binary function {:?} with input type {:?}",
+                        generic, input_type
+                    ),
+                }
+                .build()
+            })
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:391]: `with_context` instead?

File: src/flow/src/expr/func.rs [hunk 4]
```diff
@@ -154,7 +243,198 @@ pub enum BinaryFunc {
...
+            if arg_types[0].is_some() && arg_types[1].is_some() {
+                if arg_types[0] != arg_types[1] {
+                    return InvalidQuerySnafu {
+                        reason: format!(
+                            "Binary function {} requires both arguments to have the same type",
+                            name
+                        ),
+                    }
+                    .fail();
+                }
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:444]: use `ensure!` instead?

File: src/flow/src/expr/func.rs [hunk 5]
```diff
@@ -154,7 +243,198 @@ pub enum BinaryFunc {
...
+
+        if need_type && arg_type.is_none() {
+            return InvalidQuerySnafu {
+                reason: format!(
+                    "Binary function {} requires at least one argument with known type",
+                    name
+                ),
+            }
+            .fail();
+        }
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:465]: use `ensure!` instead?

File: src/flow/src/expr/func.rs [hunk 6]
```diff
@@ -154,7 +261,223 @@ pub enum BinaryFunc {
...
+                arg_types[0].clone()
+            } else if let Some(inferred) = arg_types[0].clone().or_else(|| arg_types[1].clone()) {
+                Some(inferred)
+            } else {
+                arg_exprs[0]
+                    .as_literal()
+                    .map(|lit| lit.data_type())
+                    .or_else(|| arg_exprs[1].as_literal().map(|lit| lit.data_type()))
+            }
+        };
```
Comment (zhongzc) [inline src/flow/src/expr/func.rs:451]:
```suggestion
        let arg_type = match (arg_types[0].as_ref(), arg_types[1].as_ref()) {
            (Some(t1), Some(t2)) => {
                ensure!(
                    t1 == t2,
                    InvalidQuerySnafu {
                        reason: format!(
                            "Binary function {} requires both arguments to have the same type",
                            name
                        ),
                    }
                );
                t1.clone()
            }
            (Some(t), None) | (None, Some(t)) => t.clone(),
            _ => arg_exprs[0]
                .as_literal()
                .map(|lit| lit.data_type())
                .or_else(|| arg_exprs[1].as_literal().map(|lit| lit.data_type()))
                .unwrap_or(ConcreteDataType::null_datatype()),
        };
```

File: src/flow/src/expr/relation/func.rs [hunk 7]
```diff
@@ -119,242 +129,126 @@ impl AggregateFunc {
...
+            .cloned()
+            .ok_or_else(|| {
+                InvalidQuerySnafu {
+                    reason: format!(
+                        "No specialization found for binary function {:?} with input type {:?}",
+                        generic_fn, input_type
+                    ),
+                }
+                .build()
+            })
```
Comment (zhongzc) [inline src/flow/src/expr/relation/func.rs:197]: ℹ️ `with_context`

**Review by zhongzc — APPROVED** [review]

File: src/flow/src/lib.rs [hunk 8]
```diff
@@ -12,8 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//! This crate manage dataflow in Greptime, including adapter, expr, plan, repr and utils.
+//! It can transform substrait plan into it's own plan and execute it.
+//! It also contains definition of expression, adapter and plan, and internal state management.
+
 #![allow(dead_code)]
 #![allow(unused_imports)]
+#![warn(missing_docs)]
```
Comment (waynexia) [inline src/flow/src/lib.rs:21]: Does this take the same effect like `clippy.toml`?

File: src/flow/src/expr/relation/func.rs [hunk 9]
```diff
@@ -119,242 +130,123 @@ impl AggregateFunc {
...
-    Sum,
-    Count,
-    Any,
-    All,
-}
+static SPECIALIZATION: OnceLock<HashMap<(GenericFn, ConcreteDataType), AggregateFunc>> =
+    OnceLock::new();
 
 impl AggregateFunc {
+    /// Create a `AggregateFunc`` from a string of the function name and given argument type(optional)
```
Comment (waynexia) [inline src/flow/src/expr/relation/func.rs:159]:
```suggestion
    /// Create a `AggregateFunc` from a string of the function name and given argument type(optional)
```

Comment (waynexia) [reply src/flow/src/expr/relation/func.rs:159]: Sugg: also document the `None` type param will become `Null`

File: src/flow/src/lib.rs (hunk 8 above)
Comment (discord9) [reply src/flow/src/lib.rs:21]: > Does this take the same effect like `clippy.toml`?

`missing-docs-in-crate-items` is for config `clippy::missing_docs_in_private_items`  so to only check for missing documentation in items visible within the current crate. So it should not be a overlapping with `missing_docs`
see: https://rust-lang.github.io/rust-clippy/master/index.html#/clippy::missing_docs_in_private_items

**Review by tisonkun — APPROVED** [review]

**Comment by killme2008** [comment]
@discord9 There is a compile error.

**Comment by discord9** [comment]
wired, will fix it soon

**Pull request merged**